### PR TITLE
Feature/gcp planning

### PR DESCRIPTION
**Pull request by zhuyuecai**
In this MR, we implemented the following static graph encoder based on pytorch and DGL.
They are message passing GCN(GraphConv), GAT(GATConve), SGCN(the one use DAD instead of common message passing) and GraphSage(SageConv).

For GAT, since the official release of DGL does not include GATv2Conv, we are still using the GATConv. 
Currently only GCN supported batched training. SGCN and GraphSage support mini-batch training with the sampler to sample only 1 layer. 
Still need to add mini-batch support to GAT.

A LSTM sequential decoder is added for experiment purpose.

Two examples are added in examples/ to show how to use nineturn to load ogb citation graph data and perform DTDG learning.

After adding the mini-batch support to GAT, we are good to merge this MR and close the GCN and GAT encoder planning task.

The tensorflow support for encoders is scheduled for the next MR.

### CODE REVIEW

File: nineturn/dtdg/dataloader.py [hunk 1]
```diff
@@ -52,10 +52,16 @@ def preprocess_citation_graph(graph_data: Dict[str, ndarray], node_time: str) ->
...
-    logger.debug("source shape")
-    logger.debug(sources.shape)
-    dest = np.array([nodes_id_dict[s] for s in graph_data[EDGES][1]])
+    num_edges = len(graph_data[EDGES][0])
+    tem_sources = []
+    tem_dest = []
+    for i in range(num_edges):
+        n1 = nodes_id_dict[graph_data[EDGES][0][i]]
+        n2 = nodes_id_dict[graph_data[EDGES][1][i]]
+        tem_sources.append(max(n1, n2))
```
Comment (zhuyuecai) [inline nineturn/dtdg/dataloader.py:61]: the reason for this check is because the ogb data is not clean. The source node sometimes has earlier publication year than the dst node (source cite dst) which does not make sense. So adding this to guarantee only node with later year could cite the paper with earlier year.

Comment (zhuyuecai) [reply nineturn/dtdg/dataloader.py:61]: The reason not create the list as np.array directly, because we don't want to set the array dtype. If the dtype is not good enough, say np.int8 is too small for a very large graph, then the user will be screwed up. So just leave this to python and let it determine the right dtype for the array.

File: nineturn/dtdg/models/decoder/torch/sequentialDecoder.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        # node_embs: [|V|, |hidden_dim|]
+        # sequence length = 1
+        # the sequential model processes each node at each step
+        # each snapshot the input number of nodes will change, but should only be increasing for V invariant DTDG.
+        node_embs, ids = in_state
+        if not self.mini_batch:
+            node_embs = node_embs[ids]
+        out_sequential, (h, c) = self.lstm(node_embs.view(-1, 1, self.input_d), self.memory.get_memory(ids))
+        out = self.linear(out_sequential)
+        self.memory.update_memory(h.detach(), c.detach(), ids)
```
Comment (zhuyuecai) [inline nineturn/dtdg/models/decoder/torch/sequentialDecoder.py:85]: Detach is very critical for pytorch. if your variable or tensor does not need to be BP, detach it from the compute graph, otherwise, the training will fail.

**Comment by zhuyuecai** [comment]
Merge now to facilitate the Sequential decoder design.
